### PR TITLE
add BatchStudy class

### LINKED ISSUE #906: Comparison class
I might be useful to provide a BatchStudy class or adapt the Simulation class to make it easier to compare models / discretizations / solvers / run batches of experiments and generate statistics for solvers. Usage would be something like: 

```python
batch = pybamm.BatchStudy(models={"SPM": spm, "DFN": dfn}, solvers={"casadi": casadi, "klu": klu"}, experiments={"constant current": constant_current, "cccv": cccv}, repeats=10)  #repeat solves to generate speed statistics for solvers

batch.solve()
batch.save("batch_study.p")
```

```python
batch = pickle.load("batch_study.p")
batch.plot(models=["SPM"], experiments=["cccv"], variables=["Terminal voltage [V]"])
batch.solver_stats(models=["SPM"], experiments=["constant current"]) # no argument compares all solvers
batch.export_data(models=["SPM", "DFN"], experiments=["constant current"], variables=["Terminal voltage [V]"])
```

Would be best to ensure that the class is picklable so that people can basically just batch run a whole bunch of different experiments etc on a bunch of models and solvers and then load up one object and play around with plotting / solving statistics.

### PR DESCRIPTION
**Pull request by priyanshuone6**
# Description

Add class to compare different models and conditions

Fixes #906 

## Type of change

Please add a line in the relevant section of [CHANGELOG.md](https://github.com/pybamm-team/PyBaMM/blob/develop/CHANGELOG.md) to document the change (include PR #) - note reverse order of PR #s. If necessary, also add to the list of breaking changes.

- [x] New feature (non-breaking change which adds functionality)
- [ ] Optimization (back-end change that speeds up the code)
- [ ] Bug fix (non-breaking change which fixes an issue)

# Key checklist:

- [x] No style issues: `$ flake8`
- [x] All tests pass: `$ python run-tests.py --unit`
- [x] The documentation builds: `$ cd docs` and then `$ make clean; make html`

You can run all three at once, using `$ python run-tests.py --quick`.

## Further checks:

- [x] Code is commented, particularly in hard-to-understand areas
- [x] Tests added that prove fix is effective or that feature works

### CODE REVIEW

**Comment by priyanshuone6** [comment]
What does `solver_stats` and `export_data` do?

Could you please check it? You can run this code here
https://colab.research.google.com/drive/109KJSumPxzxi3OToN6N8g6wjuD5Yb2oe?usp=sharing

File: pybamm/batch_study.py [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+            else:
+                for values in product(
+                    self.models.values(),
+                    self.solvers.values(),
+                    self.experiments.values(),
+                ):
+                    model, solver, experiment = values
+                    sim = pybamm.Simulation(model, solver=solver, experiment=experiment)
+                    sim.solve([0, 3600])
+                    self.sims.append(sim)
```
Comment (tinosulzer) [inline pybamm/batch_study.py:45]: repeats should only be called on sim.solve, and the solve times should be recorded and averaged, e.g.:
```python3
def solve(self, t_eval=None):
        self.sims = []
        if self.permutations is False:
            iter = zip(self.models.values(), self.solvers.values(), self.experiments.values())
        else:
            iter = product(self.models.values(), self.solvers.values(), self.experiments.values())
        for model, solver, experiment in zip(self.models.values(), self.solvers.values(), self.experiments.values()):
            sim = pybamm.Simulation(model, solver=solver, experiment=experiment)
            # repeat to get average solve time and integration time
            solve_time = 0
            integration_time = 0
            for num in range(self.repeats):
              sol = sim.solve(t_eval)
              solve_time += sol.solve_time
              integration_time += sol.integration_time
            sim.solution.solve_time = solve_time / self.repeats
            sim.solution.integration_time = integration_time / self.repeats
            self.sims.append(sim)
```

**Review by tinosulzer — COMMENTED** [review]
Thanks @priyanshuone6 , on the right track. See comments, and needs a unit test

**Comment by priyanshuone6** [comment]
Thanks @tinosulzer 
I am working on the unit tests, will add them soon.

**Comment by priyanshuone6** [comment]
Thanks @tinosulzer 
I am working on the unit tests, will add them soon. I have also fixed few other things I found and pushed them in the same PR rather than making a new one, hope you don't mind 😃

**Comment by priyanshuone6** [comment]
@tinosulzer The macOS and windows checks are getting cancelled as they are running for more than 6 hours, I don't know why. I ran the batch_study tests locally and I successfully got this 
```
Add -v for more debug output
..
----------------------------------------------------------------------
Ran 2 tests in 72.316s

OK
```

**Review by tinosulzer — APPROVED** [review]
Thanks @priyanshuone6 , can you just add the other options for the Simulation? parameter_values, var_pts, etc
Also add the options for `solve`

**Pull request merged**